### PR TITLE
Entity cache: link cache and authorization

### PR DESCRIPTION
**Pull request by Geal**
This builds upon #4208 to use subgraph level authorization status in  the cache key, to split cache entries per authorization contexts

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [ ] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

### CODE REVIEW

**Review by garypen — APPROVED** [review]
approving modulo: ben's comments.

File: apollo-router/tests/redis_test.rs [hunk 1]
```diff
@@ -423,4 +425,326 @@ mod test {
...
+        let response = supergraph
+            .clone()
+            .oneshot(request)
+            .await?
+            .next_response()
+            .await
+            .unwrap();
+        insta::assert_json_snapshot!(response);
+
+        let s:String = client
```
[the remark this answers is not in the document]
Comment (Geal) [reply apollo-router/tests/redis_test.rs:625]: [5812db2](https://github.com/apollographql/router/pull/4276/commits/5812db2ae2a3ad35df7203c2c15bece728aafd8d)

File: apollo-router/tests/redis_test.rs [hunk 2]
```diff
@@ -423,4 +425,326 @@ mod test {
...
+        assert_eq!(
+            v.as_object().unwrap().get("data").unwrap(),
+            &json! {{
+                "reviews": [
+                    {"body": "I can sit on it"}
+                ]
+            }}
+        );
+
+        //insta::assert_json_snapshot!(v.as_object().unwrap().get("data").unwrap());
```
[the remark this answers is not in the document]
Comment (Geal) [reply apollo-router/tests/redis_test.rs]: [5812db2](https://github.com/apollographql/router/pull/4276/commits/5812db2ae2a3ad35df7203c2c15bece728aafd8d)

File: apollo-router/tests/redis_test.rs [hunk 3]
```diff
@@ -423,4 +425,326 @@ mod test {
...
+            v.as_object().unwrap().get("data").unwrap(),
+            &json! {{
+                "reviews": [{
+                    "body": "I can sit on it",
+                    "author": {"__typename": "User", "id": "1"}
+                }]
+            }}
+        );
+
+        println!("\n\n///////////////////////////////////\n\n");
```
[the remark this answers is not in the document]
Comment (Geal) [reply apollo-router/tests/redis_test.rs]: [5812db2](https://github.com/apollographql/router/pull/4276/commits/5812db2ae2a3ad35df7203c2c15bece728aafd8d)

File: apollo-router/tests/redis_test.rs [hunk 4]
```diff
@@ -423,4 +425,326 @@ mod test {
...
+            .next_response()
+            .await
+            .unwrap();
+        insta::assert_json_snapshot!(response);
+
+        let s:String = client
+          .get("subgraph.reviews|Product|4911f7a9dbad8a47b8900d65547503a2f3c0359f65c0bc5652ad9b9843281f66|dc8e1fb584d7ad114b3e836a5fe4f642732b82eb39bb8d6dff000d844d0e3baf|f1d914240cfd0c60d5388f3f2d2ae00b5f1e2400ef2c9320252439f354515ce9")
+          .await
+          .unwrap();
+        println!("got res: {s}");
```
[the remark this answers is not in the document]
Comment (Geal) [reply apollo-router/tests/redis_test.rs]: [5812db2](https://github.com/apollographql/router/pull/4276/commits/5812db2ae2a3ad35df7203c2c15bece728aafd8d)

File: apollo-router/tests/redis_test.rs [hunk 5]
```diff
@@ -423,4 +425,326 @@ mod test {
...
+            &json! {{
+                "reviews": [
+                    {"body": "I can sit on it"}
+                ]
+            }}
+        );
+
+        //insta::assert_json_snapshot!(v.as_object().unwrap().get("data").unwrap());
+
+        println!("\n\n///////////////////////////////////\n\n");
```
[the remark this answers is not in the document]
Comment (Geal) [reply apollo-router/tests/redis_test.rs]: [5812db2](https://github.com/apollographql/router/pull/4276/commits/5812db2ae2a3ad35df7203c2c15bece728aafd8d)

File: apollo-router/tests/redis_test.rs [hunk 6]
```diff
@@ -423,4 +425,326 @@ mod test {
...
+                    "name": "table"
+                }]
+            }}
+        );
+
+        let s:String = client
+        .get("subgraph.reviews|Product|4911f7a9dbad8a47b8900d65547503a2f3c0359f65c0bc5652ad9b9843281f66|98424704ece0e377929efa619bce2cbd5246281199c72a0902da863270f5839c|d9d84a3c7ffc27b0190a671212f3740e5b8478e84e23825830e97822e25cf05c")
+        .await
+        .unwrap();
+        println!("got res: {s}");
```
[the remark this answers is not in the document]
Comment (Geal) [reply apollo-router/tests/redis_test.rs]: [5812db2](https://github.com/apollographql/router/pull/4276/commits/5812db2ae2a3ad35df7203c2c15bece728aafd8d)

File: apollo-router/tests/redis_test.rs [hunk 7]
```diff
@@ -423,4 +425,326 @@ mod test {
...
+            .next_response()
+            .await
+            .unwrap();
+        insta::assert_json_snapshot!(response);
+
+        let s:String = client
+          .get("subgraph.products|530d594c46b838e725b87d64fd6384b82f6ff14bd902b57bba9dcc34ce684b76|d9d84a3c7ffc27b0190a671212f3740e5b8478e84e23825830e97822e25cf05c")
+          .await
+          .unwrap();
+        println!("got res: {s}");
```
[the remark this answers is not in the document]
Comment (Geal) [reply apollo-router/tests/redis_test.rs]: [5812db2](https://github.com/apollographql/router/pull/4276/commits/5812db2ae2a3ad35df7203c2c15bece728aafd8d)

**Pull request merged**